yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnEventListener.scala

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history.yarn
+
+import org.apache.spark.scheduler._
+import org.apache.spark.{Logging, SparkContext}
+
+private[spark] class YarnEventListener(sc: SparkContext, service: YarnHistoryService)
+  extends SparkListener with Logging {
+
+  /**
+   * Called when a stage completes successfully or fails, with information on the completed stage.
+   */
+  override def onStageCompleted(stageCompleted: SparkListenerStageCompleted) {
+    service.enqueue(new HandleSparkEvent(stageCompleted, now()))
+  }
+
+  /**
+   * Source of current time; may be overridden in tests
+   * @return
+   */
+  protected def now(): Long = {
+    System.currentTimeMillis
+  }
+
+  /**
+   * Called when a stage is submitted
+   */
+  override def onStageSubmitted(stageSubmitted: SparkListenerStageSubmitted) {
+    service.enqueue(new HandleSparkEvent(stageSubmitted, now()))
```

Comment:
Minor, but having something like this could probably help a little:

    private def enqueue(event: SparkListenerEvent): Unit = service.enqueue(...)